peer/peer.go

```diff
@@ -1586,21 +1596,36 @@ out:
 
 		case iv := <-p.outputInvChan:
 			// No handshake?  They'll find out soon enough.
-			if p.VersionKnown() {
-				// If this is a new block, then we'll blast it
-				// out immediately, sipping the inv trickle
-				// queue.
-				if iv.Type == wire.InvTypeBlock {
-
-					invMsg := wire.NewMsgInvSizeHint(1)
-					invMsg.AddInvVect(iv)
-					waiting = queuePacket(outMsg{msg: invMsg},
-						pendingMsgs, waiting)
-				} else {
-					invSendQueue.PushBack(iv)
-				}
+			if !p.VersionKnown() {
+				continue
+			}
+
+			// If this is a new block, then we'll blast it
+			// out immediately, skipping the inv trickle
+			// queue.
+			if iv.Type == wire.InvTypeBlock {
+				invMsg := wire.NewMsgInvSizeHint(1)
+				invMsg.AddInvVect(iv)
+				waiting = queuePacket(outMsg{msg: invMsg},
+					pendingMsgs, waiting)
+				continue
+			}
+
+			// If it's a new tx and the trickle queue is enabled then enqueue the inv.
+			if useTrickleQueue {
+				invSendQueue.PushBack(iv)
+				continue
 			}
 
+			// Otherwise send it immediately
+			if p.knownInventory.Exists(iv) {
```

Comment:
Why do we need to check inventory here, and not above before we push it to the queue?

Reply:
@zquestz This check happens when consuming the queue below. I assume the reason it's checked during consumption instead of when enqueuing it is that the state of `p.knownInventory` can change so it checks as late as possible.

The check could also be done before enqueuing it but that wasn't how it was originally and I didn't see a reason to add it.